net/xia/ppal_ether/main.c

```diff
@@ -8,6 +8,83 @@
 /* ETHER_FIB table internal operations. */
 const struct xia_ppal_rt_iops *ether_rt_iops = &xia_ppal_list_rt_iops;
 
+/* Interface intialization and exit functions. */
+static struct ether_interface *eint_init(struct net_device *dev)
+{
+	struct ether_interface *eint;
+
+	ASSERT_RTNL();
+
+	eint = kzalloc(sizeof(*eint), GFP_KERNEL);
+	if (!eint)
+		return NULL;
+
+	eint->dead = 0;
+	eint->dev = dev;
+	dev_hold(dev);
+	atomic_set(&eint->refcnt, 0);
+	atomic_set(&eint->neigh_cnt, 0);
+
+	spin_lock_init(&eint->interface_lock);
+	INIT_LIST_HEAD(&eint->list_interface_common_addr);
+
+	ether_interface_hold(eint);
+	RCU_INIT_POINTER(dev->eth_ptr, eint);
+	return eint;
+}
+
+static void ether_interface_rcu_put(struct rcu_head *head)
+{
+	struct ether_interface *eint = container_of(head, struct ether_interface, rcu_head);
+	ether_interface_put(eint);
+}
+
+static void eint_destroy(struct ether_interface *eint)
+{
+	ASSERT_RTNL();
+	eint->dead = 1;
+
+	RCU_INIT_POINTER(eint->dev->eth_ptr, NULL);
+	call_rcu(&eint->rcu_head, ether_interface_rcu_put);
```

Comment:
What about if `eint->refcnt` is not zero? We should call `ether_interface_put()` here instead. Moreover, `ether_interface_put()` should, if need, call `call_rcu(...)`.